app/models/declaration/DeclarationAdditionalActorsData.scala

```diff
@@ -20,13 +20,14 @@ import forms.declaration.DeclarationAdditionalActors
 import play.api.libs.json.Json
 
 case class DeclarationAdditionalActorsData(actors: Seq[DeclarationAdditionalActors]) {
-  def containsItem(actor: DeclarationAdditionalActors): Boolean = actors.contains(actor)
+  def addActor(actor: DeclarationAdditionalActors): DeclarationAdditionalActorsData =
+    copy(if (actor.isAllowed) actors :+ actor else actors)
```

Comment:
I would recommend not to mix-in the validation here.
There is no information returned on why the operation did not succeed. I'd rather split that into two separate methods and use them together if needed (similarly to `Iterator` pattern).

Reply:
It is also validated elsewhere. I introduced this because we now have a "No" option that is sent as `partyType`. Without this, the "No" option would add a new empty actor to the collection.

Reply:
I am aware of that. This is why I suggested using two methods together. This way the method is more universal.
As it is in Iterator pattern, you have `next()` and `hasNext()` methods. And using `next()` without the other one may lead to `NullPointerException` being thrown. This design decision has been made not to break single-responsibility principle by putting both operations into single method.

Reply:
I slightly agree with @iservport .

Method addActor without validation if it's allowed doesn't make any sense, because you can't use it.

I'll rename it to `addActorIfCorrect` or something more descriptive.

Creating method to add actor to the list is like a wrapper for method inside the Sequence which do add

Reply:
I'm only not sure if this copy is necessary. I mean I don't know what is better to copy or to create a new object with new actors

Reply:
So it seems that probably the model could be better designed - not to allow adding values that are incorrect. Maybe through the use of types? But this is probably not for now.
Regarding `copy`, in my opinion it is not required, because you will copy the collection already with `:+` call. And in "else" case you are still copying the whole sequence, although it is not needed.
As a side note: is it necessary for the new actor to be appended to the sequence? It requires copying the whole sequence so it has O(n) time complexity. Prepending is O(1) and should be favoured.

Reply:
Good tip about prepending. And the copy can be avoided when `actor.isAllowed` is false as well.

Reply:
`if (actor.isAllowed) DeclarationAdditionalActorsData(actor +: actors) else this`